metadata-ingestion/src/datahub/ingestion/source/bigquery_v2/usage.py

```diff
@@ -145,6 +155,181 @@ def bigquery_audit_metadata_query_template(
+        query = f"""
+        SELECT
+            r.timestamp,
+            r.resource,
+            COUNT(q.query) query_count,
+            null as query,
+            null as query_freq,
+            null as user,
+            null as user_freq,
+            null as column,
+            null as column_freq
+        FROM
+            read_events r
+            LEFT JOIN query_events q ON r.name = q.key
+        GROUP BY r.timestamp, r.resource
+
+        UNION ALL
+
+        SELECT
+            r.timestamp,
+            r.resource,
+            null as query_count,
+            null as query,
+            null as query_freq,
+            r.user as user,
+            COUNT(r.key) as user_freq,
+            null as column,
+            null as column_freq
+        FROM
+            read_events r
+        GROUP BY r.timestamp, r.resource, r.user
+
+        UNION ALL
+
+        SELECT
+            r.timestamp,
+            r.resource,
+            null as query_count,
+            null as query,
+            null as query_freq,
+            null as user,
+            null as user_freq,
+            c.field as column,
+            COUNT(r.key) as column_freq
+        FROM
+            read_events r
+            INNER JOIN column_accesses c ON r.key = c.read_event
+        GROUP BY r.timestamp, r.resource, c.field
+
+        UNION ALL
+
+        SELECT * FROM (
+            SELECT timestamp, resource, query_count, query, query_freq, user, user_freq, column, column_freq FROM (
+                SELECT
+                    r.timestamp,
+                    r.resource,
+                    null as query_count,
+                    q.query as query,
+                    COUNT(r.key) as query_freq,
+                    ROW_NUMBER() over (PARTITION BY r.timestamp, r.resource, q.query ORDER BY COUNT(r.key) DESC, q.query) as rank,
+                    null as user,
+                    null as user_freq,
+                    null as column,
+                    null as column_freq
+                FROM
+                    read_events r
+                    LEFT JOIN query_events q ON r.name = q.key
+                GROUP BY r.timestamp, r.resource, q.query
+            ) WHERE rank <= {top_n}
+        )
+        ORDER BY
+          timestamp,
+          resource,
+          query_count DESC,
+          query_freq DESC,
+          query,
+          user_freq DESC,
+          user,
+          column_freq DESC,
+          column
+        """
+
+        rows = self.read_events.sql_query_iterator(
+            query, refs=[self.query_events, self.column_accesses]
+        )
+
+        curr: Optional[BigQueryUsageState.UsageStatistic] = None
+        for ts, res, query_count, query, q_freq, user, u_freq, column, c_freq in rows:
+            if curr is None:
+                curr = self.UsageStatistic(
+                    timestamp=ts, resource=res, query_count=query_count
+                )
+            elif (curr.timestamp, curr.resource) != (ts, res):
```

Comment:
Ah, we are relying on order by clause in SQL query here. I would like instead to push the join to sqlite to output single row with all counts, if possible.

Reply:
Looks like I can do a join. Ends up being a pretty complicated query but /seems/ to be performant. Will definitely want to test on a larger dataset before doing a full release